File: ios-app/Source/OAuth/AppAuthHandler.swift
//
// Copyright (C) 2022 Curity AB.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import AppAuth

class AppAuthHandler {
    
    private let configuration: ApplicationConfiguration
    private var userAgentSession: OIDExternalUserAgentSession?
    private var responseHandler: AuthorizationResponseHandler?
    
    init(configuration: ApplicationConfiguration) {
        self.configuration = configuration
    }

    /*
     * Get OpenID Connect endpoints
     */
    func fetchMetadata() async throws -> OIDServiceConfiguration {

        try await withCheckedThrowingContinuation { continuation in
            
            let issuerUri = URL(string: "\(self.configuration.idsvrBaseUrl)\(self.configuration.issuerPath)")!
            OIDAuthorizationService.discoverConfiguration(forIssuer: issuerUri) { metadata, ex in
                
                if metadata != nil {
                    continuation.resume(returning: metadata!)
                    
                } else {
                    
                    let error = self.createAuthorizationError(title: "Metadata Download Error", ex: ex)
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    /*
     * Trigger a redirect with standard parameters and also prompt=login
     */
    func performAuthorizationRedirect(metadata: OIDServiceConfiguration,
                                      clientID: String,
                                      viewController: UIViewController) {
        
        let redirectUriValue = URL(string: self.configuration.redirectUri)!
        
        var extraParams = [String: String]()
        extraParams["prompt"] = "login"
        
        let scopesArray = self.configuration.scope.components(separatedBy: " ")
        let request = OIDAuthorizationRequest(
            configuration: metadata,
            clientId: clientID,
            clientSecret: nil,
            scopes: scopesArray,
            redirectURL: redirectUriValue,
            responseType: OIDResponseTypeCode,
            additionalParameters: extraParams)
        
        let userAgent = OIDExternalUserAgentIOS(presenting: viewController)
        
        self.responseHandler = AuthorizationResponseHandler()
        self.userAgentSession = OIDAuthorizationService.present(
            request,
            externalUserAgent: userAgent!,
            callback: self.responseHandler!.callback)
    }

    /*
     * Finish processing, which occurs on a worker thread
     */
    func handleAuthorizationResponse() async throws -> OIDAuthorizationResponse? {

        do {
            
            return try await self.responseHandler!.waitForCallback()
            
        } catch {

            if (self.isUserCancellationErrorCode(ex: error)) {
                return nil
            }

            throw self.createAuthorizationError(title: "Authorization Request Error", ex: error)
        }
    }

    /*
     * Handle the authorization response, including the user closing the Chrome Custom Tab
     */
    func redeemCodeForTokens(authResponse: OIDAuthorizationResponse) async throws -> OIDTokenResponse {
            
        try await withCheckedThrowingContinuation { continuation in
            
            let extraParams = [String: String]()
            let request = authResponse.tokenExchangeRequest(withAdditionalParameters: extraParams)
            
            OIDAuthorizationService.perform(
                request!,
                originalAuthorizationResponse: authResponse) { tokenResponse, ex in
                    
                if tokenResponse != nil {
                    
                    continuation.resume(returning: tokenResponse!)
                    
                } else {
                    
                    let error = self.createAuthorizationError(title: "Authorization Response Error", ex: ex)
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /*
     * We can check for specific error codes to handle the user cancelling the ASWebAuthenticationSession window
     */
    private func isUserCancellationErrorCode(ex: Error) -> Bool {

        let error = ex as NSError
        return error.domain == OIDGeneralErrorDomain && error.code == OIDErrorCode.userCanceledAuthorizationFlow.rawValue
    }

    /*
     * Process standard OAuth error / error_description fields and also AppAuth error identifiers
     */
    private func createAuthorizationError(title: String, ex: Error?) -> ApplicationError {
        
        var parts = [String]()
        if (ex == nil) {

            parts.append("Authorization Error")

        } else {

            let nsError = ex! as NSError
            
            if nsError.domain.contains("org.openid.appauth") {
                parts.append("(\(nsError.domain) / \(String(nsError.code)))")
            }

            if !ex!.localizedDescription.isEmpty {
                parts.append(ex!.localizedDescription)
            }
        }

        let fullDescription = parts.joined(separator: " : ")
        let error = ApplicationError(title: title, description: fullDescription)
        return error
    }
}

File: android-app/app/src/main/java/io/curity/identityserver/client/oauth/AppAuthHandler.kt
/*
 *  Copyright 2022 Curity AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package io.curity.identityserver.client.oauth

import android.content.Context
import android.content.Intent
import io.curity.identityserver.client.configuration.ApplicationConfiguration
import io.curity.identityserver.client.errors.ApplicationException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine
import net.openid.appauth.*
import net.openid.appauth.AuthorizationServiceConfiguration.fetchFromIssuer

class AppAuthHandler(private val config: ApplicationConfiguration, val context: Context) {

    private var authService = AuthorizationService(context)

    suspend fun fetchMetadata(): AuthorizationServiceConfiguration {

        return suspendCoroutine { continuation ->

            fetchFromIssuer(this.config.getIssuerUri()) { metadata, ex ->

                when {
                    metadata != null -> {
                        continuation.resume(metadata)
                    }
                    else -> {
                        val error = createAuthorizationError("Metadata Download Error", ex)
                        continuation.resumeWithException(error)
                    }
                }
            }
        }
    }

    fun startLogin(metadata: AuthorizationServiceConfiguration, launchAction: (i: Intent) -> Unit) {

        val request = AuthorizationRequest.Builder(
            metadata,
            this.config.clientID,
            ResponseTypeValues.CODE,
            this.config.getRedirectUri())
                .setScopes(this.config.scope)
                .setPrompt("login")
                .build()

        val intent = authService.getAuthorizationRequestIntent(request)
        launchAction(intent)
    }

    suspend fun endLogin(responseIntent: Intent): TokenResponse? {

        val authorizationResponse = AuthorizationResponse.fromIntent(responseIntent)
        val ex = AuthorizationException.fromIntent(responseIntent)

        if (ex != null) {

            if (ex.type == AuthorizationException.TYPE_GENERAL_ERROR &&
                ex.code == AuthorizationException.GeneralErrors.USER_CANCELED_AUTH_FLOW.code
            ) {

                return null
            }

            throw this.createAuthorizationError("Authorization Response Error", ex)
        }

        return suspendCoroutine { continuation ->

            val extraParams = mutableMapOf<String, String>()
            val tokenRequest = authorizationResponse!!.createTokenExchangeRequest(extraParams)

            authService.performTokenRequest(tokenRequest) { tokenResponse, ex ->

                when {
                    tokenResponse != null -> {
                        continuation.resume(tokenResponse)
                    }
                    else -> {
                        val error = createAuthorizationError("Authorization Grant Error", ex)
                        continuation.resumeWithException(error)
                    }
                }
            }
        }
    }

    private fun createAuthorizationError(title: String, ex: AuthorizationException?): ApplicationException {

        val parts = mutableListOf<String>()

        if (ex?.type != null) {
            parts.add("(${ex.type} / ${ex.code})")
        }

        if (ex?.error != null) {
            parts.add(ex.error!!)
        }

        val description: String = if (ex?.errorDescription != null) {
            ex.errorDescription!!
        } else {
            "Unknown Error"
        }
        parts.add(description)

        val fullDescription = parts.joinToString(" : ")
        return ApplicationException(title, fullDescription)
    }

    fun dispose() {
        this.authService.dispose()
    }
}
